File: Leetcode/Medium/714.cpp
#include "../../headers.h"

// Time Complexity - O(N)
// Space Complexity - O(1)

class Solution
{
public:
    int maxProfit(vector<int> &p, int fee)
    {
        vector<vector<int>> dp(2, vector<int>(2, 0));
        for(int i = p.size() - 1; i >= 0; i--)
            for(int buy = 0; buy <= 1; buy++)
                if(buy)
                    dp[i % 2][buy] = max(dp[(i + 1) % 2][0] - p[i] - fee, dp[(i + 1) % 2][1]);
                else
                    dp[i % 2][buy] = max(p[i] + dp[(i + 1) % 2][1], dp[(i + 1) % 2][0]);
        return dp[0][1];
    }
};

File: 714.js
var maxProfit = function (p, fee) {
    let ans = Array(2).fill().map(() => Array(2).fill(0));
    for(let i = p.length - 1; i >= 0; i--){
        for(let buy = 0; buy <= 1; buy++){
            if(buy)
            ans[i % 2][buy] = Math.max(-fee - p[i] + ans[(i + 1) % 2][0], 0 + ans[(i + 1) % 2][0]);
            else
            ans[i % 2][buy] = Math.max(p[i] + ans[(i + 1) % 2][1], 0 + ans[(i + 1) % 2][1]);
        }
    }
    return ans[0][1];
};